// native-base-shoutem-theme/dist/index.js
Object.defineProperty(exports,"__esModule",{value:true});exports.createSharedStyle=exports.createVariations=exports.ThemeShape=exports.Theme=exports.StyleProvider=exports.INCLUDE=exports.clearThemeCache=exports.connectStyle=undefined;var _connectStyle=require("./src/connectStyle");var _connectStyle2=_interopRequireDefault(_connectStyle);
var _resolveIncludes=require("./src/resolveIncludes");
var _StyleProvider=require("./src/StyleProvider");var _StyleProvider2=_interopRequireDefault(_StyleProvider);
var _Theme=require("./src/Theme");var _Theme2=_interopRequireDefault(_Theme);
var _addons=require("./src/addons");function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}exports.


connectStyle=_connectStyle2.default;exports.
clearThemeCache=_connectStyle.clearThemeCache;exports.
INCLUDE=_resolveIncludes.INCLUDE;exports.
StyleProvider=_StyleProvider2.default;exports.
Theme=_Theme2.default;exports.
ThemeShape=_Theme.ThemeShape;exports.
createVariations=_addons.createVariations;exports.
createSharedStyle=_addons.createSharedStyle;
